### PR TITLE
Add a short path for String.concat and String.split_on_char

### PR DESCRIPTION
**Pull request by craff**
This avoid copying the string when the full string is returned.

### CODE REVIEW

File: stdlib/string.ml [hunk 1]
```diff
@@ -228,14 +229,15 @@ let hash x = seeded_hash 0 x
 (* duplicated in bytes.ml *)
 let split_on_char sep s =
   let r = ref [] in
-  let j = ref (length s) in
+  let len = length s in
+  let j = ref len in
   for i = length s - 1 downto 0 do
```
Comment (nojb) [inline stdlib/string.ml]:
```suggestion
  for i = len - 1 downto 0 do
```

File: stdlib/string.ml [hunk 2]
```diff
@@ -228,14 +229,15 @@ let hash x = seeded_hash 0 x
 (* duplicated in bytes.ml *)
```
Comment (nojb) [inline stdlib/string.ml]: Please duplicate the same change in `bytes.ml`.

**Review by nojb — APPROVED** [review]
LGTM (modulo tweaks below)

A second approval is needed. @gasche seems like a likely candidate for it.

File: stdlib/string.ml (hunk 2 above)
Comment (craff) [reply stdlib/string.ml]: Bytes.t are mutable, so we probably want to always copy the string ? Maybe the comment should be edited ?

Comment (nojb) [reply stdlib/string.ml]: Good point. In that case, I guess the comment is no longer valid and should be removed.

File: Changes [hunk 3]
```diff
@@ -71,6 +71,10 @@ Working version
   with js_of_ocaml
   (Hugo Heuzard, review by Gabriel Scherer)
 
+- #13620: Avoid copying the string in String.concat and String.split_on_char
+  when the full string is returned.
+  (Christophe Raffalli, review by ...)
```
Comment (gasche) [inline Changes]: review by Nicolás Ojeda Bär and Gabriel Scherer

**Review by gasche — APPROVED** [review]
Looks good to me -- modulo the stale comment that @nojb was pointing out. Thanks!

File: stdlib/string.ml [hunk 4]
```diff
@@ -225,17 +226,18 @@ let ends_with ~suffix s =
...
+  let len = length s in
+  let j = ref len in
+  for i = len - 1 downto 0 do
     if unsafe_get s i = sep then begin
       r := sub s (i + 1) (!j - i - 1) :: !r;
       j := i
     end
   done;
-  sub s 0 !j :: !r
+  if !j = len then [s] else sub s 0 !j :: !r
```
Comment (hhugo) [inline stdlib/string.ml]: Have you considered optimizing String.sub instead ?

File: stdlib/string.ml (hunk 2 above)
Comment (craff) [reply stdlib/string.ml]: I updated the comment, both in string.ml and bytes.ml.

File: stdlib/string.ml (hunk 4 above)
Comment (craff) [reply stdlib/string.ml]: No I did not consider this! Good idea!

Comment (craff) [reply stdlib/string.ml]: I did this, removed the special case for split_on_char and optimized String.sub. The comment is back as it is now a real duplicate.

File: stdlib/string.ml [hunk 5]
```diff
@@ -228,8 +230,9 @@ let hash x = seeded_hash 0 x
 (* duplicated in bytes.ml *)
 let split_on_char sep s =
   let r = ref [] in
-  let j = ref (length s) in
-  for i = length s - 1 downto 0 do
+  let len = length s in
+  let j = ref len in
+  for i = len - 1 downto 0 do
```
Comment (gasche) [inline stdlib/string.ml]: There is still a diff here, and in particular bytes.ml and string.ml are not identical anymore. Can you revert the change? (`length` is computed twice, but this gets eliminated by CSE anyway, so it's no big deal, and slightly less work than also rephrasing bytes.ml)

File: stdlib/string.ml [hunk 6]
```diff
@@ -40,6 +40,7 @@ let empty = ""
 let of_bytes = B.to_string
 let to_bytes = B.of_string
 let sub s ofs len =
+  if ofs = 0 && length s == len then s else
```
Comment (gasche) [inline stdlib/string.ml]: Please use `length s = len` or `Int.equal (length s) len`, see #13573.

Comment (craff) [reply stdlib/string.ml]: Too much python these days ;-) Thanks
By the way, length s = len should be enough. if ofs is not zero, it will raise invalid_arg in sub ?

Comment (gasche) [reply stdlib/string.ml]: Unless I am missing something, with `if length s = len then s else` then `String.sub s (-1) (String.length s)` will return `s` instead of failing with an error -- `B.sub` below is not called in this path. I would rather have an error in this case, so I would prefer if you also checked `ofs = 0`.

File: stdlib/string.ml [hunk 7]
```diff
@@ -40,6 +40,7 @@ let empty = ""
 let of_bytes = B.to_string
 let to_bytes = B.of_string
 let sub s ofs len =
+  if length s = len then s else
```
Comment (hhugo) [inline stdlib/string.ml]: You need to check ` ofs= 0`

Comment (gasche) [reply stdlib/string.ml]: Indeed, otherwise `sub s (-1) (String.length s` would return `s`, we want to have an error in this case. (I thought that I had written a review comment about this, but it looks like it was not posted.)

Comment (hhugo) [reply stdlib/string.ml]: It was posted here https://github.com/ocaml/ocaml/pull/13620#discussion_r1850017772

Comment (craff) [reply stdlib/string.ml]: Yes sure, that somehow works as no other value of ofs is possible, but this was a bad idea as it prevent the exception.

**Review by hhugo — APPROVED** [review]

**Comment by hhugo** [comment]
One should squash commits when merging

**Comment by craff** [comment]
> One should squash commits when merging

As I only use my laptop, for such "small" PR I work directly on github ... Hence the numerous commits. Would be nice if github add an interface to rebase -i ?

**Review by gasche — APPROVED** [review]
